### PR TITLE
TIP-694: isolate integration test files

### PR DESCRIPTION
**Pull request by jjanvier**
As a first step to make our integration tests independent, I propose, that at the beginning of each test (except if you decided to put the option `purgeDatabaseForEachTest` to false):
 - really drop the whole database schema
 - recreate the database schema 

This is just a first step, to be able to merge https://github.com/akeneo/pim-community-dev/pull/5775. With that PR, tests will be slower for sure. But this is just temporary.

What we could do after the 1.7 release is setting up a 1 day squad on this topic. Its goal would be:
- be able to launch really independent tests (with `--process-isolation` for instance, or something else). For that, we need to fix our tests.
- optimize the loading of the database which is very time consuming (see the ideas of Juju and Marie) 
- run the integration tests on Mongo
- handle EE correctly (extension points?)
    - every tests should use the PermissionsCleaner
    - schema should be totally wiped out (currently TWA needs to purge manually some tables)

(and if you're interested to work on this topic, please put a :+1: on the PR)

### CODE REVIEW

File: tests/FixturesLoader.php [hunk 1]
```diff
@@ -255,4 +258,19 @@ protected function getFilesToLoad(array $directories)
...
+        $input = new ArrayInput([
+            'command' => 'doctrine:schema:create',
+            '--env' => 'test',
+        ]);
+        $output = new BufferedOutput();
+
+        $exitCode = $this->cli->run($input, $output);
+
+        if (0 !== $exitCode) {
+            throw new \Exception(sprintf('Catalog not installable! "%s"', $output->fetch()));
```
Comment (jjanvier) [inline tests/FixturesLoader.php]: schema can not be created

**Comment by phaseinducer** [comment]
Imho you should recreate the schema in the purger, or else there will be a dependency between the purger and the loader (if you try to call the loader without having called the purger just before it will probably crash). Plus it makes more sense to me that the loader handles data only, not structure.
Also be careful: I think you'll need to modify other test cases too.
Nice PR :+1:

**Comment by jjanvier** [comment]
you're right @phaseinducer, I'll do that :)

File: tests/DatabasePurger.php [hunk 2]
```diff
@@ -14,45 +15,40 @@
  */
 class DatabasePurger
```
Comment (skeleton) [inline tests/DatabasePurger.php]: it's not really a purger now, you should rename it

**Comment by jjanvier** [comment]
@phaseinducer @skeleton I fixed your comments

File: tests/DatabaseSchemaHandler.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            'command' => 'doctrine:schema:drop',
+            '--env' => 'test',
+            '--force' => true,
+        ]);
+        $output = new BufferedOutput();
+
+        $exitCode = $this->cli->run($input, $output);
+
+        if (0 !== $exitCode) {
+            throw new \Exception(sprintf('Impossible to drop the database schema! "%s"', $output->fetch()));
```
Comment (aRn0D) [inline tests/DatabaseSchemaHandler.php]: perhaps, we should use RuntimeException ?

File: tests/DatabaseSchemaHandler.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+     * @param KernelInterface $kernel
+     */
+    public function __construct(KernelInterface $kernel)
+    {
+        $this->kernel = $kernel;
+        $this->cli = new Application($this->kernel);
+        $this->cli->setAutoExit(false);
+    }
+
+    public function reset()
```
Comment (aRn0D) [inline tests/DatabaseSchemaHandler.php]: You should add doc for each method

File: tests/DatabaseSchemaHandler.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    /** @var Application */
+    protected $cli;
+
+    /**
+     * @param KernelInterface $kernel
+     */
+    public function __construct(KernelInterface $kernel)
+    {
+        $this->kernel = $kernel;
```
Comment (aRn0D) [inline tests/DatabaseSchemaHandler.php]: This class attribute is not used, perhaps we should remove it.

File: tests/DatabaseSchemaHandler.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class DatabaseSchemaHandler
+{
+    /** @var KernelInterface */
+    protected $kernel;
+
+    /** @var Application */
+    protected $cli;
```
Comment (aRn0D) [inline tests/DatabaseSchemaHandler.php]: this should be private because we should not extends this class

File: tests/FixturesLoader.php [hunk 7]
```diff
@@ -117,27 +126,20 @@ protected function loadImportFiles(array $files)
...
+                'code'     => $jobInstance->getCode(),
+                '--no-log' => true,
+                '-v'       => true
+            ]);
+            $output = new BufferedOutput();
+            $exitCode = $this->cli->run($input, $output);
 
             if (0 !== $exitCode) {
-                throw new \Exception(sprintf('Catalog not installable! "%s"', $command->getDisplay()));
+                throw new \Exception(sprintf('Catalog not installable! "%s"', $output->fetch()));
```
Comment (aRn0D) [inline tests/FixturesLoader.php]: RuntimeException ?

File: tests/FixturesLoader.php [hunk 8]
```diff
@@ -117,27 +126,20 @@ protected function loadImportFiles(array $files)
...
-                ]
-            );
+            $input = new ArrayInput([
+                'command'  => 'akeneo:batch:job',
+                'code'     => $jobInstance->getCode(),
+                '--no-log' => true,
+                '-v'       => true
+            ]);
+            $output = new BufferedOutput();
+            $exitCode = $this->cli->run($input, $output);
```
Comment (aRn0D) [inline tests/FixturesLoader.php]: Definitely, this ways is cleaner !

File: tests/DatabaseSchemaHandler.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+use Symfony\Component\Console\Input\ArrayInput;
+use Symfony\Component\Console\Output\BufferedOutput;
+use Symfony\Component\HttpKernel\KernelInterface;
+
+/**
+ * @author    Yohan Blain <yohan.blain@akeneo.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class DatabaseSchemaHandler
```
Comment (aRn0D) [inline tests/DatabaseSchemaHandler.php]: handler sounds manager :)

**Comment by aRn0D** [comment]
Don't forget the EE edition !

File: tests/DatabaseSchemaHandler.php (hunk 6 above)
Comment (jjanvier) [inline tests/DatabaseSchemaHandler.php]: @aRn0D extended by yourself for the TWA

File: tests/DatabaseSchemaHandler.php (hunk 5 above)
Comment (jjanvier) [inline tests/DatabaseSchemaHandler.php]: used in the override...
I agree we should take care of this problem
see the description of the PR

File: tests/DatabaseSchemaHandler.php (hunk 3 above)
Comment (jjanvier) [inline tests/DatabaseSchemaHandler.php]: done

File: tests/FixturesLoader.php (hunk 7 above)
Comment (jjanvier) [inline tests/FixturesLoader.php]: done

File: tests/DatabaseSchemaHandler.php (hunk 4 above)
Comment (jjanvier) [inline tests/DatabaseSchemaHandler.php]: done

File: tests/DatabaseSchemaHandler.php (hunk 9 above)
Comment (jjanvier) [inline tests/DatabaseSchemaHandler.php]: completely agree, that sucks
maybe we could find something better during the task force?

**Pull request merged**